### DISCUSSION TITLE
Create standalone executables from mruby scripts without a C toolchain.

### DISCUSSION THREAD

**Discussion by suetanvil** [mruby/mruby#5633] [Show and tell]
I've written a tool that will package an mruby program as a standalone native executable:

```bash
$ echo 'puts "Hello, world"' >> hello.rb
$ monolith hello.rb -o hello
$ file hello
hello: Mach-O 64-bit executable x86_64
$ ./hello 
Hello, world
```

It's a gem and so can be incorporated into typical mruby projects reasonably easily.  The resulting tool depends on `mrbc` but does not require any other compiler, linker, or toolchain component to be present.

It works by taking advantage of the fact that modern operating systems will ignore data that has been appended to executables.  (This is how self-extracting zip files work, for example.)  So this gem builds a simple executable (`monolith_run`) linked to `libmruby` that finds its own executable, opens it, seeks to the start of the `.mrb` files that have been appended to it and loads them with `mrb_load_irep_file()`.

The accompanying tool (`monolith`) will compile your mruby source(s) with `mrbc` and append the result to a copy of `monolith_run`.  It is itself a "compiled" mruby script.

The gem also adds a small extension to `mruby` to help  with development.

The project is [here](https://github.com/suetanvil/mruby-bin-monolith).

**Comment by matz** [comment]
Shall I add mruby-bin-monolith to mgem-list? (Sorry for too late response)

**Reply by suetanvil** [reply]
Yes, please.

**Reply by suetanvil** [reply]
Do you need me to provide a pull request?

**Reply by matz** [reply]
I have an issue to create a pull request. Please submit a pull request by yourself.

**Reply by suetanvil** [reply]
It has been submitted.

**Reply by matz** [reply]
Thank you!
